File: xml-inflater/src/main/java/com/itsaky/inflater/IViewGroup.java
/*
 * This file is part of AndroidIDE.
 *
 * AndroidIDE is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * AndroidIDE is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with AndroidIDE.  If not, see <https://www.gnu.org/licenses/>.
 *
 */
package com.itsaky.inflater;

import android.view.View;

import java.util.List;
import java.util.function.Consumer;

/**
 * Represents a {@link android.view.ViewGroup}
 *
 * @author Akash Yadav
 */
public interface IViewGroup extends IView {

  /**
   * Add this view to this group
   *
   * @param view The view to add
   */
  void addView(IView view);

  /**
   * Add this view to this group
   *
   * @param view The view to add
   * @param index Add this view to this index
   */
  void addView(IView view, int index);

  /**
   * Remove the view at the specified index
   *
   * @param index The index of the view to remove
   */
  void removeView(int index);

  /**
   * Remove the given if it exists in this group
   *
   * @param view The view to remove
   */
  void removeView(IView view);

  /** Remove all children in this group */
  void removeAll();

  /**
   * Perform an action in every child in this group
   *
   * @param consumer The consumer which will consume the views
   */
  void forEachChild(Consumer<IView> consumer);

  /**
   * Get the number of child views this layout contains
   *
   * @return The child count
   */
  int getChildCount();

  /**
   * Get the list of children in this group
   *
   * @return The list of children
   */
  List<IView> getChildren();

  /**
   * Get child at the provided index
   *
   * @param index The index of child
   * @throws IndexOutOfBoundsException If there is no child at the given index
   */
  IView getChildAt(int index);

  /**
   * Find the index of the given children in this group.
   *
   * @param view The view to find index of.
   * @return The index of the child or -1 if this group does not contain the given child.
   */
  int indexOfChild(IView view);

  /**
   * Find the index of the given children in this group.
   *
   * @param view The view to find index of.
   * @return The index of the child or -1 if this group does not contain the given child.
   */
  int indexOfChild(View view);

  /**
   * Register the given hierarchy change listener to this group.
   *
   * @param listener The listener to register.
   */
  void registerHierarchyChangeListener(OnHierarchyChangeListener listener);

  /**
   * Unregister the given hierarchy change listener from this view group.
   *
   * @param listener The listener to unregister.
   */
  void unregisterHierarchyChangeListener(OnHierarchyChangeListener listener);

  /**
   * A listener which can be used to get notification when a children is added or removed from an
   * {@link IViewGroup}.
   *
   * @author Akash Yadav
   */
  public static interface OnHierarchyChangeListener {

    /**
     * Called when a child view is added in the view group to which this listener is attached.
     *
     * @param view The view that was added.
     */
    void onViewAdded(IView view);

    /**
     * Called when a child view is removed from the view group to which this listener is attached.
     *
     * @param view The view that was removed. Note that this view's parent will be null.
     */
    void onViewRemoved(IView view);
  }
}

File: xml-inflater/src/main/java/com/itsaky/androidide/inflater/IViewGroup.kt
/*
 *  This file is part of AndroidIDE.
 *
 *  AndroidIDE is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  AndroidIDE is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *   along with AndroidIDE.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.itsaky.androidide.inflater

/**
 * Inflated model for a [ViewGroup][android.view.ViewGroup].
 *
 * @author Akash Yadav
 */
interface IViewGroup : IView {

  /** The number of children in this view group. */
  val childCount: Int

  /**
   * Add the given child.
   *
   * @param view The view to add.
   */
  fun addChild(view: IView)

  /**
   * Add the given [view] at the given [index].
   *
   * @param index The index at which the child should be added.
   * @param view The child view to add .
   */
  fun addChild(index: Int, view: IView)

  /**
   * Remove the given child.
   *
   * @param view The child to remove.
   */
  fun removeChild(view: IView)

  /**
   * Remove the child at the given index.
   *
   * @param index The index of the child to remove.
   */
  fun removeChild(index: Int)

  /**
   * Get the child at the given index.
   *
   * @param index The index of the child.
   */
  operator fun get(index: Int): IView

  /**
   * Replace the child at the given index.
   *
   * @param index The index of the child to replace.
   * @param view The new view to set.
   * @return The existing child or `null`.
   */
  operator fun set(index: Int, view: IView): IView

  /**
   * Finds the index of the given child in this view group.
   *
   * @param child The child.
   * @return The index of the child if found or -1.
   */
  fun indexOfChild(child: IView): Int {
    for (i in 0 until childCount) {
      if (this[i].view == child.view) {
        return i
      }
    }
    return -1
  }

  /**
   * Adds the given hierarchy change listener.
   *
   * @param listener The listener to add.
   */
  fun addOnHierarchyChangeListener(listener: OnHierarchyChangeListener)

  /**
   * Removes the given hierarchy change listener.
   *
   * @param listener The listener to remove.
   */
  fun removeOnHierarchyChangeListener(listener: OnHierarchyChangeListener)

  /** Listener to listen for hierarchy changes in an [IViewGroup]. */
  interface OnHierarchyChangeListener {
  
    /**
     * Called just before the [view] is added to the [group].
     *
     * @param group The group in which the view will be added.
     * @param view The view that will be added to the given view group.
     */
    fun beforeViewAdded(group: IViewGroup, view: IView) {}
  
    /**
     * Called just before the [view] is removed from the [group].
     *
     * @param group The group from which the view will be removed.
     * @param view The view that will be removed from the given view group.
     */
    fun beforeViewRemoved(group: IViewGroup, view: IView) {}
    
    /**
     * Called when a view is added.
     *
     * @param group The view group in which the view was added.
     * @param view The added view.
     */
    fun onViewAdded(group: IViewGroup, view: IView)

    /**
     * Called when a view is removed.
     *
     * @param group The view group from which the view was removed.
     * @param view The removed view.
     */
    fun onViewRemoved(group: IViewGroup, view: IView)
  }

  /** Allows overriding a single method in [OnHierarchyChangeListener]. */
  open class SingleOnHierarchyChangeListener : OnHierarchyChangeListener {
    override fun onViewAdded(group: IViewGroup, view: IView) {}
    override fun onViewRemoved(group: IViewGroup, view: IView) {}
  }
}
